config_proc_macro/src/style_edition.rs

```diff
@@ -0,0 +1,355 @@
+    se2015: Option<syn::Expr>,
+    se2018: Option<syn::Expr>,
+    se2021: Option<syn::Expr>,
+    se2024: Option<syn::Expr>,
+}
+
+impl StyleEditionDefault {
+    /// a sinlge default for all style editions
+    fn single_default(&self) -> bool {
+        self.default.is_some()
+            && self.se2015.is_none()
+            && self.se2018.is_none()
+            && self.se2021.is_none()
+            && self.se2024.is_none()
+    }
+    /// Infer the type from the default value
+    fn ty_from_default(&self) -> syn::Result<syn::Type> {
+        match &self.default {
+            Some(syn::Expr::Lit(lit)) => match lit.lit {
+                syn::Lit::Bool(_) => {
+                    return Ok(syn::TypePath {
+                        qself: None,
+                        path: path_from_str("bool"),
+                    }
+                    .into());
+                }
+                syn::Lit::Int(_) => {
+                    return Ok(syn::TypePath {
+                        qself: None,
+                        path: path_from_str("usize"),
+                    }
+                    .into());
+                }
+                _ => {}
+            },
+            _ => {}
+        }
+        Err(syn::parse::Error::new(
+            Span::call_site(),
+            "could not determine type from default value",
+        ))
+    }
+
+    fn enum_expr_path(varient: &syn::Variant, en: &syn::ItemEnum) -> syn::ExprPath {
+        let mut path = path_from_ident(&en.ident);
+        path.segments.push(varient.ident.clone().into());
+        syn::ExprPath {
+            attrs: vec![],
+            qself: None,
+            path,
+        }
+    }
+
+    /// Set the style edition based on the the annotated attribute
+    /// For example:
+    /// ```ignore
+    /// #[style_edition]
+    /// enum Example {
+    ///     #[se_default] // <-- Default style edition
+    ///     A,
+    ///     #[se_2018] // <-- Explicit override for StypeEdition2018
+    ///     B,
+    /// }
+    /// ```
+    fn set_defaults_by_enum_variant_attr(&mut self, en: &syn::ItemEnum) {
+        for varient in en.variants.iter() {
+            for attr in varient.attrs.iter() {
+                if se_default(attr) {
+                    self.default.replace(Self::enum_expr_path(varient, en).into());
+                    break;
+                } else if se_2015(attr) {
+                    self.se2015.replace(Self::enum_expr_path(varient, en).into());
+                    break;
+                } else if se_2018(attr) {
+                    self.se2018.replace(Self::enum_expr_path(varient, en).into());
+                    break;
+                } else if se_2021(attr) {
+                    self.se2021.replace(Self::enum_expr_path(varient, en).into());
+                    break;
+                } else if se_2024(attr) {
+                    self.se2024.replace(Self::enum_expr_path(varient, en).into());
+                    break;
+                }
```

Comment:
This doesn't check whether there are two attributes, which we probably want to error? (i.e. something like `#[se_2018] #[se_2021]`)

Also, I would prefer to replace this by getting the attr path as an ident, getting the ident as a string and then matching on the string, probably via let-else and continue if we can't unwrap it?